ui/raidboss/data/04-sb/eureka/eureka_hydatos.ts

```diff
@@ -515,6 +525,7 @@ const triggerSet: TriggerSet<Data> = {
       // In other words, they need to be the fully translated in game log.
       // There are no "gainsEffect" lines for the clones, only for Absolute Virtue directly.
       // Ideally parser logic could be added for this case, but this is where we are.
+      // TODO: Possibly animation data from ActorControlExtra that controls the light/dark wrists?
```

Comment:
I think that these effect gains are available now, they were part of the StatusEffectList2 and/or StatusEffectList3 packets which weren't available at the time? But I haven't done the AV fight so I'm not sure.

Reply:
> I think that these effect gains are available now, they were part of the StatusEffectList2 and/or StatusEffectList3 packets which weren't available at the time? But I haven't done the AV fight so I'm not sure.

This one was very odd.  I looked through an AV log (thanks @xiashtra !) and it didn't have the `GameLog` lines that these triggers use (and chat log was not hidden).  So I couldn't see how these triggers would have fired, even in current state.

I'll spend a little more time with the log to see if there's an option for using status effects instead.  But testing may be tough to come by - I'm not likely to jump back into BA anytime soon.

Reply:
> I'll spend a little more time with the log to see if there's an option for using status effects instead. But testing may be tough to come by - I'm not likely to jump back into BA anytime soon.

I'm not seeing any 0x1A lines that could be used as suitable replacements; there are some instances where the boss self-applies `6AE` (Astral Essence) and `6AF` (Umbral Essence), but none of those seem to line up with triple impact stream.  If we were going to replace this trigger, I think we'd need to (a) capture the animation being used by each clone, and then (b) determine the clone firing order.  As for that order, the current logic uses chat lines, but from the log I'm looking at, I don't see those chat lines anywhere, so I can't see how those triggers function now.

We could look at the `StartsUsing` order of the Relative Virtue clones coupled with their animation state... but the position data in this log for the RV actors doesn't look right to me.  The first-firing clone is at (-163,298) and the second is at (-163,330).  It seems strange that both would have the same `x` value, because you're supposed to use their position around the arena (and the light-dark split relative to their position) to figure out which clone is firing next. 
If they are 180 degrees apart, I can't see how you'd figure out which is about to fire, which has me wondering if these RV actors are really the right ones to be looking at.

tldr; this is going to take some runs with logs/vod to unpack, and since I'm unlikely to be back in BA anytime soon, this will likely have to stay an open TODO for now.

Reply:
> > I think that these effect gains are available now, they were part of the StatusEffectList2 and/or StatusEffectList3 packets which weren't available at the time? But I haven't done the AV fight so I'm not sure.
> 
> This one was very odd. I looked through an AV log (thanks @xiashtra !) and it didn't have the `GameLog` lines that these triggers use (and chat log was not hidden). So I couldn't see how these triggers would have fired, even in current state.
> 
> I'll spend a little more time with the log to see if there's an option for using status effects instead. But testing may be tough to come by - I'm not likely to jump back into BA anytime soon.

I looked through some other BA logs I have saved and found a few that do have the lines from these triggers, but they are only in ~1/3 of the BA logs I have and they are not from a consistent time period.  I'm going to try and get into a few BA runs and see if I can investigate more why they are not being written consistently.

Reply:
Per Discord conversation with @wexxlee, it seems that the regex on these triggers isn't correct; they aren't firing due to special characters present in the chatline:
```00|2024-03-17T22:44:49.0000000-04:00|332E||Relative Virtue gains the effect of Umbral Essence.|98acce60df5f0a61
00|2024-03-17T22:44:49.0000000-04:00|332E||Relative Virtue gains the effect of Umbral Essence.|cd0cc61b9cc9881f
00|2024-03-17T22:44:50.0000000-04:00|332E||Relative Virtue gains the effect of Astral Essence.|48e5f0b52de318d1```
Furthermore, it appears that these chatlines aren't enabled by the default XIV Log Filters; I had to enable the options to show `Beneficial effects granted by unengaged enemies.` and `Beneficial effects granted to unengaged enemies.` to have them written to the log in the first place.

Reply:
I'll clean up the regex to account for the unicode, but if these log lines are buried behind a game setting (that's off by default), we *really* should replace these triggers.

Disregard my earlier comment about the position data being wonky.  The clones were on intercards, hence the same x value (because, yeah, that's how circles work).  In any case, we can determine clone firing order based on the order of each Relative Virtue combatant's 0x14 line for `3795` (Impact Stream).

What we don't have a solve for right now is determining their effect/animation to indicate which color is safe.  There's no 0x1A, 0x26, or 0x2A lines for the `6AE` Astral Essence/`6AF` Umbral Essence effects.

@valarnin - since we've confirmed these chat lines are (or at least can be) written, is it possible there's a `StatusEffectList` packet, but no line emitted from ACT because the combatant is invisible/unengaged when the effect is being applied?  If so, maybe this is justification for a `StatusListExtra` or similar OP line that would provide status effect data for these combatants (assuming of course that is the issue)?

Reply:
If someone's got a log with packet capture enabled, I could check. Otherwise I'll have to get back in to Eureka and run it myself to check.

I don't even know where AV is tbh, I did a single run of Eureka and never saw it so I'm guessing maybe it's in the support group side or something?

Reply:
> If someone's got a log with packet capture enabled, I could check. Otherwise I'll have to get back in to Eureka and run it myself to check.
> 
> I don't even know where AV is tbh, I did a single run of BA and never saw it so I'm guessing maybe it's in the support group side or something?

AV is the 3rd Boss inside BA just before the Final Boss staircase (where you have to wait for the support group to finish the special fate outside before entering the fight to weaken AV)

Reply:
Hmm, maybe the day I ran BA was the day my ACT crashed and I didn't notice... I'll have to get another run in to check I guess.

Reply:
@xiashtra, if you're headed back in again sometime soon, think you could enable packet capture for AV?

Reply:
```
26|2024-03-18T19:11:44.1650000-04:00|6AF|Umbral Essence|9999.00|400027B0|Absolute Virtue|400027B0|Absolute Virtue|00|5858587|5858587|678ca4361787d1c2
```
???

I'm seeing the status effect in opcode `641` (0x281) for these actors (not posting the full lines since it has player IDs in it, in reverse byte order, so they're annoying to filter out), timestamped `2024-03-18T19:11:44.1740000-04:00`:
```
40002650
40002652
40002653
40002651
40002654
40002655
40002657
40002659
40002656
40002658
4000265A
```

But the chat line for this happens before either of the actual network packets (raw packet or the `26` line) get dumped to file? And opcode `641` doesn't line up with any of the opcodes in <https://github.com/ravahn/machina/blob/master/Machina.FFXIV/Headers/Opcodes/Global.txt>, I would have expected that to be the opcode for `StatusEffectList3`...

Looking even further into this, I think that the status effect is actually encoded on the `21` line for the `Eidos` cast itself here, in the `Damage` field?
```
21|2024-03-18T19:11:43.5460000-04:00|400027B0|Absolute Virtue|3787|Eidos|400027B0|Absolute Virtue|E|6AF0000|1B|37878000|0|0|0|0|0|0|0|0|0|0|0|0|4699508|5858587|10000|10000|||-189.90|313.80|82.00|-1.56|4699508|5858587|10000|10000|||-189.90|313.80|82.00|-1.56|0001AA58|0|1|ad8433ddd5799e0e
flags = 0xE
damage high = 0x06AF
damage low = 0x0000
```

Later on with the 3 clones, it looks like they only gain the buff via ActorControl packets, so ACT never emits a normal log line for those? (this was a different run, ignore the large difference in timestamp to previous examples)

```
4096|2024-03-18T20:54:39.6610000-04:00|ActorControl|40005B89|GainEffect|000006AF|00000000|00000000|00000000|00000000|05e43a2eaa7cb1c5
4096|2024-03-18T20:54:39.6610000-04:00|ActorControl|40005B8C|GainEffect|000006AE|00000000|00000000|00000000|00000000|3fb5d121f87c880a
4096|2024-03-18T20:54:39.6610000-04:00|ActorControl|40005B8F|GainEffect|000006AE|00000000|00000000|00000000|00000000|755704f2396d4ea0

00|2024-03-18T20:54:39.0000000-04:00|332E||Relative Virtue gains the effect of Umbral Essence.|1d1901a5943f9f31
00|2024-03-18T20:54:39.0000000-04:00|332E||Relative Virtue gains the effect of Astral Essence.|28053d2aebea695a
00|2024-03-18T20:54:39.0000000-04:00|332E||Relative Virtue gains the effect of Astral Essence.|b0cdcaa506042cca
```

Unfortunately all 6 instances of `Eidos` I have from today were Umbral Essence so I can't confirm that Astral will have the correct ID in the `Eidos` cast.

Reply:
> ???
> 
> I'm seeing the status effect in opcode `641` (0x281) for these actors (not posting the full lines since it has player IDs in it, in reverse byte order, so they're annoying to filter out), timestamped `2024-03-18T19:11:44.1740000-04:00`:
> [...]
> 
> Looking even further into this, I think that the status effect is actually encoded on the `21` line for the `Eidos` cast itself here, in the `Damage` field?

I might not be fully tracking your thoughts here, but I don't think the status effect on Absolute Virtue (or AV hidden combatants) is necessarily the data point needed to resolve the mechanic?  There is at least one (maybe multiple? can't recall) instances where the main boss does Eidos and then a single Impact Stream, and I think there are hidden AV combatants to handle the vfx/damage.

From looking at the data for Triple Impact Stream, though, it's the 3 Relative Virtue Clones with the glowing wrists (`6AE`/`6AF`) that determine which half of the arena is unsafe.  At the same position as each RV actor, there are also two hidden AV actors, each facing 90 degrees away from the RV actor's heading (which should be toward the middle of the arena), each one doing either Bright Aurora or Dark Aurora on the half they are facing.  But knowing that (or which buff they have) wouldn't, I don't think, tell us which half will be safe?  (Sorry if I'm misunderstanding your point.)

> Later on with the 3 clones, it looks like they only gain the buff via ActorControl packets, so ACT never emits a normal log line for those? (this was a different run, ignore the large difference in timestamp to previous examples)
> 
> ```
> 4096|2024-03-18T20:54:39.6610000-04:00|ActorControl|40005B89|GainEffect|000006AF|00000000|00000000|00000000|00000000|05e43a2eaa7cb1c5
> 4096|2024-03-18T20:54:39.6610000-04:00|ActorControl|40005B8C|GainEffect|000006AE|00000000|00000000|00000000|00000000|3fb5d121f87c880a
> 4096|2024-03-18T20:54:39.6610000-04:00|ActorControl|40005B8F|GainEffect|000006AE|00000000|00000000|00000000|00000000|755704f2396d4ea0
> 
> 00|2024-03-18T20:54:39.0000000-04:00|332E||Relative Virtue gains the effect of Umbral Essence.|1d1901a5943f9f31
> 00|2024-03-18T20:54:39.0000000-04:00|332E||Relative Virtue gains the effect of Astral Essence.|28053d2aebea695a
> 00|2024-03-18T20:54:39.0000000-04:00|332E||Relative Virtue gains the effect of Astral Essence.|b0cdcaa506042cca
> ```

Yep, I think those are the packets we need.  Since each one of those hidden actors will then cast `Impact Stream` in sequence (cast time of 4.7 seconds, with a little wiggle room in between for damage application), if we've pre-collected this packet data, we should be able to call out whether Light or Dark is safe for each cast.

Looking at the ActorControl volume table you put together, `GainEffect` is up there, relatively speaking, in terms of "noisy" categories.  And this seems like a fairly niche case, so reasonable minds could differ on whether it's worth it to add it to `ActorControlExtra`.  But I don't think there's any other way to resolve this mechanic with the data we have, so I'd be in favor of adding it.  (I suppose there's always the option to add this category just for this particular zone, but .... yeah, not sure how I feel about OP filtering the lines it emits based on zone.)

Reply:
> I have an open question to Ravahn in the dev channel of the discord about the `Relative Virtue` case.

I'm planning to merge with a todo/comment - but assuming we don't end up with a definitive answer, do you have thoughts on whether it would be worthwhile to add this to `ActorControlExtra` (and if/how we should filter to prevent duplication of existing 0x1A lines?  Perhaps just these two effects?)